### PR TITLE
add juju config to use custom images for knative serving

### PR DESCRIPTION
**Pull request by ca-scribner**
This adds to knative-serving:
* the `custom_images` config to to the charm, allowing users to provide overrides to the default images so they can use their own (for example, if they work in an air-gapped environment)
* `DEFAULT_IMAGES` defined in the charm that allow the charm to have an opinion on what images should be used if none are specified
* tests for the above features

### Manual test instructions

```
cd charms/knative-serving
charmcraft pack
juju deploy knative-operator --trust --channel latest/edge
juju deploy ./knative-serving_ubuntu-20.04-amd64.charm --trust --config namespace=knative-serving
kubectl get pod -n knative-serving  # observe the knative serving pods
juju config knative-serving custom_images="{activator: not-a-real-image}"
kubectl get pod -n knative-serving  # observe how a new activator pod is started.  This will not deploy successfully as the image above is not real
juju config knative-serving custom_images=""  # reset back to defaults
kubectl get pod -n knative-serving  # observe the knative serving pods
```

### CODE REVIEW

**Comment by misohu** [comment]
Amazing work Andrew .... I would add test for wrong config (to test that the charm goes to blocked state) <3

**Review by misohu — COMMENTED** [review]
Amazing work Andrew .... I would add test for wrong config (to test that the charm goes to blocked state) <3

File: charms/knative-serving/src/charm.py
Comment (DnPlas) [inline charms/knative-serving/src/charm.py:150]: Just to keep the context property a bit tidy, I wonder if this could be done outside of the `_context` property and just update the context dict with this information, similar to what is done for `self._otel_collector_relation_data`.

File: charms/knative-serving/src/image_management.py
Comment (DnPlas) [inline charms/knative-serving/src/image_management.py:15]: I'm okay with raising `ValueError` here as, according to the [docs](https://docs.python.org/3/library/exceptions.html#ValueError): "Raised when an operation or function receives an argument that has the right type but an inappropriate value, and the situation is not described by a more precise exception such as [IndexError](https://docs.python.org/3/library/exceptions.html#IndexError).", so I could agree that if the `yaml` is not formatted correctly we could say it is a value error when `YAMLError` is raised, but it took me a while to understand why this is the case. Probably it would be more clear if the docstring of this method stated in which situation a `yaml.YAMLError` is raised and why we want to raise back a `ValueError` from that other exception.

Also, this is completely optional, but I believe `YAMLError` has information about the line and column where it found the parsing error.

File: charms/knative-serving/src/image_management.py
Comment (DnPlas) [inline charms/knative-serving/src/image_management.py:29]: This docstring should say "returns" somewhere.

File: charms/knative-serving/tests/unit/test_charm.py
Comment (DnPlas) [inline charms/knative-serving/tests/unit/test_charm.py:209]:
```suggestion
    Note: This test is trivial now, where custom_image_config always equals custom_images,
    but ...
```

Comment (DnPlas) [reply charms/knative-serving/tests/unit/test_charm.py:209]: I'm not entirely sure what you mean with this note, does this mean this implementation will change in the near future?

File: charms/knative-serving/src/image_management.py
Comment (DnPlas) [inline charms/knative-serving/src/image_management.py:16]: Since this message is passed all the way to the user through the blocked status message, should we add instructions on how to move forward to unblock the unit?
If not here, then we should tell users how to unblock the unit in the charm code.

File: charms/knative-serving/tests/unit/test_charm.py
Comment (DnPlas) [inline charms/knative-serving/tests/unit/test_charm.py:220]: I'd add test case that asserts we are actually going into `BlockedStatus` when `ValueError` is raised.

File: tests/test_bundle.py
Comment (DnPlas) [inline tests/test_bundle.py:10]: We should probably add this as a requirement in `knative-operators/requirements-integration.in`, as we are explicitly calling it in the test.py file.

File: tests/test_bundle.py
Comment (DnPlas) [inline tests/test_bundle.py:215]: Are we restoring the default settings in the case we add another test case that requires a working `knative-serving`?

**Review by DnPlas — COMMENTED** [review]
Thanks @ca-scribner ! Left a couple of comments, but looks good overall.

File: charms/knative-serving/README.md
Comment (ca-scribner) [reply charms/knative-serving/README.md]: This better?  This is also the knative-serving README.md (not the general one for the repo)

File: charms/knative-serving/README.md
Comment (ca-scribner) [reply charms/knative-serving/README.md:30]: Tried rewording - does this work better?

File: charms/knative-serving/config.yaml
Comment (ca-scribner) [reply charms/knative-serving/config.yaml:37]: Added!

File: charms/knative-serving/requirements-integration.in
Comment (ca-scribner) [reply charms/knative-serving/requirements-integration.in:5]: Good catch.  It was for the repo-level integration.txt, but I accidentally added it to both.  I've removed `serving/*integration.*` because, as you say, it is unused.

File: charms/knative-serving/src/image_management.py
Comment (ca-scribner) [reply charms/knative-serving/src/image_management.py:15]: yeah I wasn't sure what was best here.  I wanted to simplify/add context for the error so the message could be better caught for a status, but it didn't quite feel right either.  

I've updated the docstrings, but give another read and see if it works for you.  I'm open to suggestions.  I like this as is if we're also logging the full error message (including the YAMLError, which like you say has more specific info).  But if we just catch to set a status, it feels opaque

File: charms/knative-serving/tests/unit/test_charm.py
Comment (ca-scribner) [reply charms/knative-serving/tests/unit/test_charm.py:209]: yeah I think when we have rocks for this, this dict will be where they go.  For now we have no custom images, so we just use upstream's

Comment (ca-scribner) [reply charms/knative-serving/tests/unit/test_charm.py:209]: Tried rewording a little, but not sure its better

File: charms/knative-serving/src/image_management.py
Comment (ca-scribner) [reply charms/knative-serving/src/image_management.py:16]: agreed!  added to the blocked message.

File: charms/knative-serving/src/image_management.py
Comment (ca-scribner) [reply charms/knative-serving/src/image_management.py:15]: Maybe I should just let the YAMLError go uncaught, and just catch it in the charm and make a blocked message there instead.  wdyt?

File: tests/test_bundle.py
Comment (ca-scribner) [reply tests/test_bundle.py:215]: Exactly

File: charms/knative-serving/tests/unit/test_charm.py
Comment (ca-scribner) [reply charms/knative-serving/tests/unit/test_charm.py:220]: Added `test_custom_images_config_context_with_incorrect_config` to handle this

File: charms/knative-serving/src/image_management.py
Comment (DnPlas) [reply charms/knative-serving/src/image_management.py:19]: Yeah, I think that last option is good.

File: tests/test_bundle.py
Comment (ca-scribner) [reply tests/test_bundle.py:10]: added

File: charms/knative-serving/src/image_management.py
Comment (ca-scribner) [reply charms/knative-serving/src/image_management.py:19]: Updated.  wdyt?

File: charms/knative-serving/src/image_management.py
Comment (DnPlas) [inline charms/knative-serving/src/image_management.py:27]: Please change the docstring to reflect the return of this method, if possible also add its return type

**Review by DnPlas — COMMENTED** [review]
Just two tiny comments, other than that LGTM.

File: charms/knative-serving/src/charm.py
Comment (ca-scribner) [reply charms/knative-serving/src/charm.py]: updated!

File: charms/knative-serving/src/image_management.py
Comment (ca-scribner) [reply charms/knative-serving/src/image_management.py]: updated!

File: requirements-integration.in
Comment (DnPlas) [inline requirements-integration.in:6]:
```suggestion
pytest-asyncio
```

**Review by DnPlas — DISMISSED** [review]
Thanks @ca-scribner, just a minor nit, please make sure you fix it before merging.

**Review by DnPlas — APPROVED** [review]

**Pull request merged**
